# NanoPlot
Plotting scripts for nanopore sequencing data

![Example plot](https://github.com/wdecoster/NanoPlot/blob/master/examples/scaled_Log_Downsampled_LengthvsQualityScatterPlot_kde.png)

The example plot above shows a bivariate plot comparing log transformed read length with average basecall Phred quality score.

More examples can be found in the [gallery on my blog 'Gigabase Or Gigabyte'.](https://gigabaseorgigabyte.wordpress.com/2017/06/01/example-gallery-of-nanoplot/)

This script performs data extraction from Nanopore sequencing data in the following formats:
- fastq files (can be bgzip, bzip2 or gzip compressed)  
- fastq files generated by albacore containing additional information (can be bgzip, bzip2 or gzip compressed)  
- bam files  



Various plots are created automatically, more can be created optionally.

The script is written for python3 but might also work for python2.7 (untested).

### USAGE:
```
usage: NanoPlot.py [-h] [--threads THREADS] [--time] [--maxlength MAXLENGTH]
                   [--drop_outliers] [--downsample DOWNSAMPLE] [--loglength]
                   [--alength] [--outdir OUTDIR] [--version] [--prefix PREFIX]
                   (--fastq FASTQ | --fastq_albacore FASTQ_ALBACORE | --bam BAM)

Required argument is one of these:
--fastq FASTQ         Data presented is in fastq format.
--fastq_albacore FASTQ_ALBACORE
                      Data presented is in fastq format generated by
                      albacore with additional information concerning
                      channel and time.
--bam BAM             Data presented as a bam file.

Optional arguments:
  -h, --help            show this help message and exit
  --threads THREADS     Set the allowed number of threads to be used by the
                        script
  --maxlength MAXLENGTH
                        Drop reads longer than length specified.
  --drop_outliers       Drop outlier reads with extreme long length.
  --downsample DOWNSAMPLE
                        Reduce dataset to N reads by random sampling.
  --loglength           Logarithmic scaling of lengths in plots.
  --alength             Use aligned read lengths rather than sequenced length
                        (bam mode)
  --outdir OUTDIR       Specify directory in which output has to be created.
  --version             Print version and exit.
  --prefix PREFIX       Specify an optional prefix to be used for the output
                        files.
```

## Companion script NanoFilt.py
Script to perform filtering on quality and/or read length, and optional trimming after passing filters
Reads from stdin, writes to stdout
Intended to be used:
- directly after fastq extraction
- prior to mapping
- in a stream between extraction and mapping

### USAGE:
```
usage: NanoFilt.py [-h] [-q QUALITY] [-l LENGTH] [--headcrop HEADCROP] [--tailcrop TAILCROP]

optional arguments:  
  -h, --help            show this help message and exit  
  -q QUALITY, --quality QUALITY  Filter on a minimum average read quality score  
  -l LENGTH, --length LENGTH Filter on a minimum read length  
  --headcrop HEADCROP   Trim n nucleotides from start of read  
  --tailcrop TAILCROP   Trim n nucleotides from end of read
```
